File: internal/api/servergrpc/auditlogs.go
package servergrpc

import (
	"context"

	serverv1 "github.com/torchwoodcloud/torchwood/genproto/server/v1"
	sharedv1 "github.com/torchwoodcloud/torchwood/genproto/shared/v1"
	appserver "github.com/torchwoodcloud/torchwood/internal/app/server"
	"github.com/torchwoodcloud/torchwood/internal/domain/audit"
	"github.com/torchwoodcloud/torchwood/pkg/crud"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
	"google.golang.org/protobuf/types/known/structpb"
	"google.golang.org/protobuf/types/known/timestamppb"
)

type AuditLogsService struct {
	serverv1.UnimplementedAuditLogsServiceServer
	auditLogs *appserver.AuditLogs
}

func NewAuditLogsService(auditLogs *appserver.AuditLogs) *AuditLogsService {
	return &AuditLogsService{auditLogs: auditLogs}
}

// ListAuditLogs 查询审计日志：offset 分页（pkg/crud token，仿 apikeys），
// 结构化过滤 exact 匹配，项目作用域语义在 app 用例层（决策对齐 Outbox）。
func (s *AuditLogsService) ListAuditLogs(ctx context.Context, req *serverv1.ListAuditLogsRequest) (*serverv1.ListAuditLogsResponse, error) {
	params, err := crud.ParseListParams(req.GetPageSize(), req.GetPageToken(), "", "")
	if err != nil {
		return nil, status.Error(codes.InvalidArgument, err.Error())
	}
	query := appserver.AuditLogsQuery{
		ActorID:         req.GetActorId(),
		ActorKind:       req.GetActorKind(),
		Action:          req.GetAction(),
		Status:          req.GetStatus(),
		ResourceID:      req.GetResourceId(),
		IncludePlatform: req.GetIncludePlatform(),
		AllProjects:     req.GetAllProjects(),
		Offset:          params.Offset,
		PageSize:        int(params.PageSize),
	}
	if ts := req.GetCreatedAfter(); ts != nil {
		t := ts.AsTime()
		query.CreatedAfter = &t
	}
	if ts := req.GetCreatedBefore(); ts != nil {
		t := ts.AsTime()
		query.CreatedBefore = &t
	}
	entries, total, err := s.auditLogs.List(ctx, query)
	if err != nil {
		return nil, err
	}
	info := crud.BuildPaginationInfo(params, total, params.Offset+int(params.PageSize) < total)
	var nextToken, prevToken string
	if info.HasNext {
		if nextToken, err = crud.EncodePageToken(info.NextOffset); err != nil {
			return nil, status.Error(codes.Internal, err.Error())
		}
	}
	if info.HasPrevious {
		if prevToken, err = crud.EncodePageToken(info.PreviousOffset); err != nil {
			return nil, status.Error(codes.Internal, err.Error())
		}
	}
	out := make([]*serverv1.AuditLog, len(entries))
	for i := range entries {
		out[i] = mapAuditLog(&entries[i])
	}
	return &serverv1.ListAuditLogsResponse{
		AuditLogs: out,
		Meta: &sharedv1.ListResponseMeta{
			PageSize:      info.PageSize,
			TotalCount:    int32(info.TotalCount),
			NextPageToken: nextToken,
			PrevPageToken: prevToken,
		},
	}, nil
}

func mapAuditLog(e *audit.Entry) *serverv1.AuditLog {
	out := &serverv1.AuditLog{
		Id:         e.ID,
		ProjectId:  e.ProjectID,
		ActorId:    e.ActorID,
		ActorKind:  e.ActorKind,
		Action:     e.Action,
		Status:     e.Status,
		ResourceId: e.ResourceID,
		Ip:         e.IP,
		UserAgent:  e.UserAgent,
		CreatedAt:  timestamppb.New(e.CreatedAt),
	}
	if e.Metadata != nil {
		if md, err := structpb.NewStruct(e.Metadata); err == nil {
			out.Metadata = md
		}
	}
	return out
}

File: console/src/api/auditLogs.ts
import { api } from "./client";

// 字段为 snake_case：grpc-gateway 以 json_names_for_fields=false 序列化
// （与 users.ts 等现有封装一致）。
export interface AuditLog {
  id: string;
  project_id?: string;
  actor_id?: string;
  actor_kind?: string;
  action: string;
  status: string;
  resource_id?: string;
  ip?: string;
  user_agent?: string;
  metadata?: Record<string, unknown>;
  created_at?: string;
}

export interface ListAuditLogsParams {
  page_size?: number;
  page_token?: string;
  action?: string;
  status?: string;
  actor_id?: string;
  resource_id?: string;
  created_after?: string;
  created_before?: string;
  include_platform?: boolean;
  all_projects?: boolean;
}

export interface ListAuditLogsResponse {
  audit_logs: AuditLog[];
  meta?: {
    total_count?: number;
    page_size?: number;
    next_page_token?: string;
    prev_page_token?: string;
  };
}

// listAuditLogs 查询审计日志（项目上下文经 client 拦截器注入
// X-Torchwood-Project；include_platform/all_projects 仅平台 admin 可用，
// 服务端 PermissionDenied 兜底）。
export async function listAuditLogs(
  params: ListAuditLogsParams
): Promise<ListAuditLogsResponse> {
  const clean: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(params)) {
    if (v !== undefined && v !== "" && v !== false) {
      clean[k] = v;
    }
  }
  const res = await api.get<ListAuditLogsResponse>("/server/audit-logs", {
    params: clean,
  });
  return {
    audit_logs: res.data.audit_logs ?? [],
    meta: res.data.meta,
  };
}
